R/compare.R

```diff
@@ -0,0 +1,274 @@
+           ifelse(files$hash.x == files$hash.y, "unchanged", "modified")))
+  data.frame(path = files$path, status = status)
+}
+
+compare_files <- function(target, current, files, root, search_options) {
+  path_target <- withr::local_tempdir()
+  path_current <- withr::local_tempdir()
+
+  # Copying the files into a temporary directory is fairly wasteful and, as
+  # long as the packet is unpacked already, we could read the files from the
+  # archive or file store directly. Nevertheless this makes accessing the files
+  # very straightforward, and covers the case where the file only exists
+  # remotely transparent.
+
+  orderly_copy_files(target, dest = path_target, files = files,
+                     options = search_options, root = root)
+
+  orderly_copy_files(current, dest = path_current, files = files,
+                     options = search_options, root = root)
+
+  ret <- lapply(files, function(p) {
+    if (is_binary_file(file.path(path_target, p)) ||
+        is_binary_file(file.path(path_current, p))) {
+      NULL
+    } else {
+      diffobj::diffChr(
+        read_file_lossy(file.path(path_target, p)),
+        read_file_lossy(file.path(path_current, p)),
+        tar.banner = file.path(target, p),
+        cur.banner = file.path(current, p),
+        rds = FALSE,
+        mode = "unified",
+        style = list(pad = FALSE, wrap = FALSE),
+        interactive = FALSE)
+    }
+  })
+  names(ret) <- files
+
+  ret
+}
+
+##' Compare the metadata and contents of two packets.
+##'
+##' Insignificant differences in the metadata (eg. different dates and packet
+##' IDs) are excluded from the comparison.
+##'
+##' If either packet is not unpacked, but `search_options`'s `allow_remote` is
+##' `TRUE`, we will try to request files from remote locations, as necessary.
+##'
+##' @title Compare two packets
+##'
+##' @param target The id of the packet to use in the comparison.
+##' @param current The id of the other packet against which to compare.
+##' @param what One or more of "metadata", "files" and "artefacts", retricting
+##'   what components of the packet to compare. This is useful when it is known
+##'   for example that the source code of a report what changed, and one is only
+##'   interested in the effect on its output.
+##' @param search_options Options for locating packet files. If there are no
+##'   copies of the files locally, they can be downloaded automatically from a
+##'   remote location on-demand if `allow_remote` is `TRUE`.
+##'
+##' @inheritParams orderly_metadata
+##'
+##' @return An R6 object of class `orderly_packet_diff` is returned. Printing
+##'   this object will show the difference in the two packets. Additionally, the
+##'   `is_equal` method of the object return a logical indicating whether any
+##'   difference was found.
+##'
+##' @export
+orderly_compare_packets <- function(
+  target, current, search_options = NULL, root = NULL, locate = TRUE,
+  what = c("metadata", "files")) {
```

Comment:
I wonder if we should accept `verbose` here somehow? I can't immediately think of a way though and you may have already thought about this.  What I tried was:

```
> files <- orderly_compare_packets(p1, p2, what = "files", root = root)
> files
ℹ Comparing packets 20240919-155913-1eb3a847 and 20240919-155913-573ea930
! The following files exist in both packets but have different contents:
• data.R
ℹ Print the comparison with `verbose = TRUE` to display the differences in the files' contents
> files <- orderly_compare_packets(p1, p2, what = "files", root = root, verbose = TRUE)
Error in orderly_compare_packets(p1, p2, what = "files", root = root,  : 
  unused argument (verbose = TRUE)
```

Reply:
I guess it's odd to have `what` here, and not as part of the print method, given `verbose` is there.  Can we do the comparison at this point, and shift `what` there?